### DISCUSSION TITLE
Is it possible to convert to tensorrt?

### DISCUSSION THREAD

**Discussion by catalwaysright** [openai/whisper#169] [Q&A]
Since I noticed that it can be exported to onnx, I wonder if it supports tensorrt conversion.

**Comment by Y-T-G** [comment]
I was able to convert the encoder to TensorRT without any issues. But I have trouble converting the decoder to TensorRT.

**Reply by catalwaysright** [reply]
What problem did you meet?

**Reply by Y-T-G** [reply]
@catalwaysright The output from the compiled model is incorrect and I don't know know how to handle the dynamic input and output shapes in the inference code.

**Reply by catalwaysright** [reply]
@Y-T-G Hi! Sorry for the late reply. I think I know how to handle the dynamic input and output shape, so maybe your engine is fine. Can we get in touch and could you share your script for converting? We can see what is actually going on.

**Reply by Y-T-G** [reply]
@catalwaysright I uploaded the script [here](https://github.com/Y-T-G/whisper-tensorrt/blob/dev/compile_trt.py). I used [this code](https://github.com/zhuzilin/whisper-openvino/blob/edb69443a541955b849a3743d567910eb023597a/whisper/decoding.py#L153) to export the decoder to ONNX. You can contact me on Telegram if you wish: 
![image](https://user-images.githubusercontent.com/32206511/196288010-4014d052-064c-4665-82f0-8e9cba6ef38b.png)

**Reply by ejkitchen** [reply]
Were you able to get it to work finally?

**Reply by Y-T-G** [reply]
@ejkitchen No. I still haven't figured out how to compile it correctly with dynamic shapes that's required by the inference.

**Reply by catalwaysright** [reply]
> @ejkitchen No. I still haven't figured out how to compile it correctly with dynamic shapes that's required by the inference.

I think I am making some progress, but I have some problems with the kv_cache in decoder.

**Reply by ejkitchen** [reply]
I have someone else converting to Pytorch script and TensorRT. Will also share code once completed but if someone finishes first, we can share ideas!

**Reply by Anjum48** [reply]
@Y-T-G , @catalwaysright @ejkitchen were you guys able to get TensorRT to work? I'm new to TensorRT but I'm assuming the encoder & decoder will need to be separate models?

**Reply by Y-T-G** [reply]
@Anjum48 Yes, they have to be separate. Someone else already made a TensorRT compatible version: https://github.com/FamousDirector/FastWhisper

**Reply by Anjum48** [reply]
I've spent a bit of time with the FastWhisper implementation and I believe that there are some issues with the implementation. 

- The sample audio is 4 seconds long, which gets padded to 30 seconds (87% silence). This is ok though, we can use longer audio
- Switching to a longer audio sample makes the ONNX conversion time for the tiny model from about ~1 minute to 40 minutes on my machine. This fixes the issue where the ONNX model couldn't reproduce the results of the regular (greedy) model in the Gradio demo when a longer sample is selected
- Why does the conversion take so much longer? There is a `while True` loop in the forward pass that is unwrapped in the graph representation. For the 4s clip, this was broken after ~400 iterations (hence why the ONNX model couldn't transcribe longer audio), whereas for a full clip this will be 3000 iterations (the time dim of the Mel spectrogram), allowing longer transcriptions
- For ONNX this is ok, but causes considerable issues for TensorRT. I'm not sure, but I believe that after unwrapping these 3000 iterations TensorRT tries to optimize the graph for each iteration individually since it has no way of knowing that part of the graph was a loop. I have had a `trtexec` session running for > 36 hours now consuming 40GB of RAM set on max verbosity with no output for the last 24 hours. I think that the graph has become too big for TensorRT to really optimize properly

I did find this [example](https://github.com/NVIDIA/TensorRT/tree/main/demo/HuggingFace#how-to-run-with-k-v-cache) where a HF BART model was converted using TensorRT by separating the encoder & decoder while also preserving the KV cache. I think something similar needs to be done for Whisper

I also tried converting the `decoder.onnx` files from the Openvino implementation but I got a `This version of TensorRT does not support dynamic axes` error. I'm pretty sure that dynamic axes should be supported though, so I'll look into this

**Reply by Y-T-G** [reply]
@FamousDirector

**Reply by Y-T-G** [reply]
@Anjum48 Why do you need to use a long audio for conversion? Wouldn't the result be the same? 

The while loop is in the forward pass of the Whisper model which is a wrapper that utilizes the encoder and decoder model. What needs to be converted to TensorRT are just the encoder (AudioEncoder) and decoder (TextDecoder) models. I don't see how longer audio sample would affect the outcome of the final TensorRT or ONNX models.

**Reply by catalwaysright** [reply]
I have succeeded in converting Whisper to tensorrt by getting rid of the kv_cache. It works fine. However, the inference speed is actually slower the the pytorch model. Thus, I am confused about the point why I converted into tensorrt lol..

**Reply by Y-T-G** [reply]
@catalwaysright Without kv_cache, it will be slower because it is doing the same redundant computation multiple times.

**Reply by catalwaysright** [reply]
@Y-T-G I modified the pytorch model too. Neither of them use kv_cache right now so it is the pure compare of inference speed.

**Reply by Y-T-G** [reply]
@Anjum48 There's another complicated way of doing the conversion using KV cache [here](https://github.com/ELS-RD/transformer-deploy/blob/main/demo/generative-model/t5.ipynb). @pommedeterresautee explains how he did it fo T5.

**Reply by Y-T-G** [reply]
@catalwaysright I see.

I am looking forward to Kernl support for Whisper. https://github.com/ELS-RD/kernl/pull/122

**Reply by Anjum48** [reply]
> Why do you need to use a long audio for conversion? Wouldn't the result be the same?

The results are not the same. Steps to reproduce:

1. Create an ONNX model with the 4-sec sample clip.
2. Transcribe a clip longer than 4 seconds in the Gradio app and note the results
3. Create an ONNX model using an input clip > 30 seconds.
4. Repeat step 2

You'll see that the transcription from step 2 is truncated to the first few seconds, whereas step 4 has the full 30-second transcription. My suspicion is that the difference is caused by the while loop breaking as soon as there is silence/padding during graph creation, so the full clip is not processed.

**Reply by Y-T-G** [reply]
@Anjum48 I think the ONNX export script needs to be modified. It's exporting the whole Whisper model. You could try splitting it up to export the encoder and decoder part separately. That way the loop is not involved in the graph. cc: @FamousDirector

**Reply by FamousDirector** [reply]
Glad to see the FastWhisper repo is generating some discussion. Its implementation is certainly **not** production ready and should only serve as some general guidance. 

To answer some questions:
- Yes, the decoder and encoder could be individually separated and converted, then used in tandem. Originally, I wanted it to be a single engine hence what I did. You'll likely want to compile the Python code or move it to C++ to maintain a higher speed. Also, may want to look at data on the GPU to reduce inter-device communication time.
- The original Whisper implementation also only does 30 sec at a time. See here: https://github.com/openai/whisper/blob/main/whisper/transcribe.py#L174 - you could implement something similar with a single engine of smaller length, feeding in the previous tokens, thus not needing as long as an input
